bazaarvoice/emodb
# Added adaptive CQL result sets to retry when appropriate with smaller fetch sizes

## Github Issue #

[60](https://github.com/bazaarvoice/emodb/issues/60)

## What Are We Doing Here?

As described in the corresponding issue EmoDB uses a version of the Datastax CQL driver which contains an extremely harmful bug when a query returns more than 256MB.  In short, when the maximum frame size is exceeded the driver effectively closes multiple connections to Cassandra and renders them unusable.  Over time this results in a complete failure of the Emo service once all connections enough servers have been closed.

This pull request has two changes:

1. Upgrades to version 3.1.1 of the Datastax CQL driver which does not contain the defect.
2. Catches exceptions for frame size exceptions and automatically retries the query with a smaller fetch size.

## How to Test and Verify

1. Check out this PR
2. Build Emo with the command:
   `mvn -DrunAdaptiveCqlIntegrationTests=true clean install`
    There is a unit test specifically for this error, but since it takes a long time it does not run unless the Java parameter above is included.
3. Regression test to verify existing functionality is unchanged.

## Risk

Because CQL reads are at the heart of almost all Emo functionality the risk is that any errors introduced here could have wide reaching consequences.  A slow, methodological rollout to production is recommended.

### Level 

`Medium` Although the risks introduced by this change are high the change itself is fairly straightforward and doesn't significantly change the existing query infrastructure.  Error handling takes place at the `ResultSet` level so all of the more complicated logic concerning row grouping, resolution, compaction, caching and reloading is unchanged.

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
